File: libs/document-processor-next/src/engine/processor-engine.ts
import { Blue, BlueNode } from '@blue-labs/language';

import { ChannelRunner, type ChannelMatch } from './channel-runner.js';
import { CheckpointManager } from './checkpoint-manager.js';
import type {
  ChannelBinding,
  ContractBundle,
  HandlerBinding,
} from './contract-bundle.js';
import type { ContractLoader } from './contract-loader.js';
import { ProcessorExecutionContext } from './processor-execution-context.js';
import { ScopeExecutor } from './scope-executor.js';
import { TerminationService } from './termination-service.js';
import type { ExecutionAdapter } from './processor-execution-context.js';
import type { TerminationExecutionAdapter } from './termination-service.js';
import {
  DocumentProcessingRuntime as Runtime,
  type DocumentProcessingRuntime,
  type DocumentUpdateData,
} from '../runtime/document-processing-runtime.js';
import type { JsonPatch } from '../model/shared/json-patch.js';
import { canonicalSignature } from '../util/node-canonicalizer.js';
import {
  normalizePointer,
  normalizeScope,
  relativizePointer,
  resolvePointer,
} from '../util/pointer-utils.js';
import type { Node } from '../types/index.js';
import type { ContractProcessorRegistry } from '../registry/contract-processor-registry.js';
import type { ChannelContract } from '../model/index.js';
import type { ChannelEvaluationContext } from '../registry/types.js';
import type { TerminationKind } from '../runtime/scope-runtime-context.js';
import { ProcessorErrors } from '../types/errors.js';
import { DocumentProcessingResult } from '../types/document-processing-result.js';
import { RunTerminationError } from './run-termination-error.js';
import { ProcessorFatalError } from './processor-fatal-error.js';
import { MustUnderstandFailure } from './must-understand-failure.js';
import { IllegalStateException } from './illegal-state-exception.js';
import { z } from 'zod';

interface ExecutionHooks extends ExecutionAdapter, TerminationExecutionAdapter {
  bundleForScope(scopePath: string): ContractBundle | undefined;
  deliverLifecycle(
    scopePath: string,
    bundle: ContractBundle | null,
    event: Node,
    finalizeAfter: boolean
  ): void;
}

export class ProcessorExecution implements ExecutionHooks {
  private readonly runtimeRef: Runtime;
  private readonly bundles = new Map<string, ContractBundle>();
  private readonly pendingTerminations = new Map<
    string,
    { kind: TerminationKind; reason: string | null }
  >();
  private readonly cutOffScopes = new Set<string>();
  private readonly checkpointManager: CheckpointManager;
  private readonly terminationService: TerminationService;
  private readonly channelRunner: ChannelRunner;
  private readonly scopeExecutor: ScopeExecutor;
  private readonly blue: Blue;

  constructor(
    private readonly contractLoader: ContractLoader,
    private readonly registry: ContractProcessorRegistry,
    blue: Blue,
    document: Node
  ) {
    this.runtimeRef = new Runtime(document);
    this.blue = blue;
    this.checkpointManager = new CheckpointManager(
      this.runtimeRef,
      canonicalSignature
    );
    this.terminationService = new TerminationService(this.runtimeRef);
    this.channelRunner = new ChannelRunner(
      this.runtimeRef,
      this.checkpointManager,
      {
        evaluateChannel: (channel, bundle, scopePath, event) =>
          this.evaluateChannel(channel, bundle, scopePath, event),
        isScopeInactive: (scopePath) => this.isScopeInactive(scopePath),
        createContext: (scopePath, bundle, event, allowTerminatedWork) =>
          this.createContext(
            scopePath,
            bundle,
            event,
            allowTerminatedWork,
            false
          ),
        executeHandler: (handler, context) =>
          this.executeHandler(handler, context),
      }
    );
    this.scopeExecutor = new ScopeExecutor({
      runtime: this.runtimeRef,
      contractLoader: this.contractLoader,
      channelRunner: this.channelRunner,
      bundles: this.bundles,
      hooks: {
        isScopeInactive: (scopePath) => this.isScopeInactive(scopePath),
        createContext: (
          scopePath,
          bundle,
          event,
          allowTerminatedWork,
          allowReserved
        ) =>
          this.createContext(
            scopePath,
            bundle,
            event,
            allowTerminatedWork,
            allowReserved ?? false
          ),
        recordLifecycleForBridging: (scopePath, event) =>
          this.recordLifecycleForBridging(scopePath, event),
        enterFatalTermination: (scope, bundle, reason) =>
          this.enterFatalTermination(scope, bundle, reason ?? null),
        fatalReason: (error, label) =>
          (error instanceof Error && error.message) || label,
        markCutOff: (scopePath) => this.markCutOff(scopePath),
      },
      blueId: (node) => this.blue.calculateBlueIdSync(node),
      nodeAt: (scopePath) => this.nodeAt(scopePath),
      createDocumentUpdateEvent: (data, scopePath) =>
        this.createDocumentUpdateEvent(data, scopePath),
      matchesDocumentUpdate: (scopePath, watchPath, changedPath) =>
        this.matchesDocumentUpdate(scopePath, watchPath, changedPath),
    });
  }

  initializeScope(scopePath: string, chargeScopeEntry: boolean): void {
    this.scopeExecutor.initializeScope(scopePath, chargeScopeEntry);
  }

  loadBundles(scopePath: string): void {
    this.scopeExecutor.loadBundles(scopePath);
  }

  processExternalEvent(scopePath: string, event: Node): void {
    this.scopeExecutor.processExternalEvent(scopePath, event);
  }

  handlePatch(
    scopePath: string,
    bundle: ContractBundle,
    patch: JsonPatch,
    allowReservedMutation: boolean
  ): void {
    this.scopeExecutor.handlePatch(
      scopePath,
      bundle,
      patch,
      allowReservedMutation
    );
  }

  createContext(
    scopePath: string,
    bundle: ContractBundle,
    event: Node,
    allowTerminatedWork = false,
    allowReservedMutation = false
  ): ProcessorExecutionContext {
    return new ProcessorExecutionContext(
      this,
      bundle,
      scopePath,
      event.clone(),
      allowTerminatedWork,
      allowReservedMutation,
      this.blue
    );
  }

  result(): DocumentProcessingResult {
    const document = (this.runtimeRef.document() as BlueNode).clone();
    const triggeredEvents = this.runtimeRef
      .rootEmissions()
      .map((event) => (event as BlueNode).clone());
    return DocumentProcessingResult.of(
      document,
      triggeredEvents as readonly BlueNode[],
      this.runtimeRef.totalGas()
    );
  }

  runtime(): DocumentProcessingRuntime {
    return this.runtimeRef;
  }

  bundleForScope(scopePath: string): ContractBundle | undefined {
    return this.bundles.get(normalizeScope(scopePath));
  }

  isScopeInactive(scopePath: string): boolean {
    const normalized = normalizeScope(scopePath);
    return (
      this.cutOffScopes.has(normalized) ||
      this.pendingTerminations.has(normalized) ||
      this.runtimeRef.isScopeTerminated(normalized)
    );
  }

  enterGracefulTermination(
    scopePath: string,
    bundle: ContractBundle | null,
    reason: string | null
  ): void {
    this.terminate(scopePath, bundle, 'GRACEFUL', reason);
  }

  enterFatalTermination(
    scopePath: string,
    bundle: ContractBundle | null,
    reason: string | null
  ): void {
    this.terminate(scopePath, bundle, 'FATAL', reason);
  }

  recordPendingTermination(
    scopePath: string,
    kind: TerminationKind,
    reason: string | null
  ): void {
    this.pendingTerminations.set(normalizeScope(scopePath), { kind, reason });
  }

  clearPendingTermination(scopePath: string): void {
    this.pendingTerminations.delete(normalizeScope(scopePath));
  }

  markCutOff(scopePath: string): void {
    const normalized = normalizeScope(scopePath);
    if (this.cutOffScopes.add(normalized)) {
      const context = this.runtimeRef.existingScope(normalized);
      context?.markCutOff();
    }
  }

  deliverLifecycle(
    scopePath: string,
    bundle: ContractBundle | null,
    event: Node,
    finalizeAfter: boolean
  ): void {
    this.scopeExecutor.deliverLifecycle(
      scopePath,
      bundle,
      event,
      finalizeAfter
    );
  }

  recordLifecycleForBridging(scopePath: string, event: Node): void {
    const context = this.runtimeRef.scope(scopePath);
    context.recordBridgeable(event.clone());
    if (scopePath === '/') {
      this.runtimeRef.recordRootEmission(event.clone());
    }
  }

  normalizeScope(scopePath: string): string {
    return normalizeScope(scopePath);
  }

  resolvePointer(scopePath: string, relativePointer: string): string {
    return resolvePointer(scopePath, relativePointer);
  }

  private terminate(
    scopePath: string,
    bundle: ContractBundle | null,
    kind: TerminationKind,
    reason: string | null
  ): void {
    const normalized = normalizeScope(scopePath);
    if (
      this.pendingTerminations.has(normalized) ||
      this.runtimeRef.isScopeTerminated(normalized)
    ) {
      return;
    }
    this.pendingTerminations.set(normalized, { kind, reason });
    this.terminationService.terminateScope(
      this,
      scopePath,
      bundle,
      kind,
      reason
    );
  }

  private nodeAt(scopePath: string): Node | null {
    const normalized = normalizeScope(scopePath);
    return ProcessorEngine.nodeAt(this.runtimeRef.document(), normalized);
  }

  private evaluateChannel(
    channel: ChannelBinding,
    bundle: ContractBundle,
    scopePath: string,
    event: Node
  ): ChannelMatch {
    const processor = this.registry.lookupChannel(channel.blueId());
    if (!processor) {
      return { matches: false };
    }

    const eventClone = event.clone();
    let eventObject: ChannelEvaluationContext['eventObject'] = null;
    try {
      eventObject = this.blue.nodeToSchemaOutput(eventClone, z.object({}));
    } catch {
      eventObject = null;
    }
    const evaluationContext: ChannelEvaluationContext = {
      scopePath,
      blue: this.blue,
      event: eventClone,
      eventObject,
      markers: bundle.markers(),
    };

    const matchesResult = processor.matches(
      channel.contract() as ChannelContract,
      evaluationContext
    );
    if (matchesResult instanceof Promise) {
      throw new Error('Async channel processors are not supported');
    }
    if (!matchesResult) {
      return { matches: false };
    }

    const eventIdResult = processor.eventId?.(
      channel.contract() as ChannelContract,
      evaluationContext
    );
    if (eventIdResult instanceof Promise) {
      throw new Error('Async channel processors are not supported');
    }

    return {
      matches: true,
      eventId: eventIdResult ?? null,
      eventNode: eventClone.clone(),
    };
  }

  private executeHandler(
    handler: HandlerBinding,
    context: ProcessorExecutionContext
  ): void {
    const processor = this.registry.lookupHandler(handler.blueId());
    if (!processor) {
      const reason = `No processor registered for handler contract ${handler.blueId()}`;
      throw new ProcessorFatalError(
        reason,
        ProcessorErrors.illegalState(reason)
      );
    }
    processor.execute(handler.contract(), context);
  }

  private createDocumentUpdateEvent(
    data: DocumentUpdateData,
    scopePath: string
  ): Node {
    const relativePath = relativizePointer(scopePath, data.path);

    const beforeNode =
      data.before != null ? data.before.clone() : new BlueNode().setValue(null);
    const afterNode =
      data.after != null ? data.after.clone() : new BlueNode().setValue(null);

    const eventNode = new BlueNode();
    eventNode.setProperties({
      type: new BlueNode().setValue('Document Update'),
      op: new BlueNode().setValue(data.op),
      path: new BlueNode().setValue(relativePath),
      before: beforeNode,
      after: afterNode,
    });
    return eventNode;
  }

  private matchesDocumentUpdate(
    scopePath: string,
    watchPath: string | null | undefined,
    changedPath: string
  ): boolean {
    if (!watchPath || watchPath.length === 0) {
      return false;
    }
    const watch = normalizePointer(resolvePointer(scopePath, watchPath));
    const changed = normalizePointer(changedPath);
    if (watch === '/') {
      return true;
    }
    if (changed === watch) {
      return true;
    }
    return changed.startsWith(`${watch}/`);
  }
}

export class ProcessorEngine {
  constructor(
    private readonly contractLoader: ContractLoader,
    private readonly registry: ContractProcessorRegistry,
    private readonly blue: Blue
  ) {}

  initializeDocument(document: Node): DocumentProcessingResult {
    if (this.isInitialized(document)) {
      throw new IllegalStateException('Document already initialized');
    }
    const execution = this.createExecution(document.clone());
    return this.run(document, execution, () => {
      execution.initializeScope('/', true);
    });
  }

  processDocument(document: Node, event: Node): DocumentProcessingResult {
    if (!this.isInitialized(document)) {
      throw new IllegalStateException('Document not initialized');
    }
    const execution = this.createExecution(document.clone());
    const eventClone = event.clone();
    return this.run(document, execution, () => {
      execution.loadBundles('/');
      execution.processExternalEvent('/', eventClone);
    });
  }

  isInitialized(document: Node): boolean {
    return this.initializationMarker(document) != null;
  }

  createExecution(document: Node): ProcessorExecution {
    return new ProcessorExecution(
      this.contractLoader,
      this.registry,
      this.blue,
      document
    );
  }

  private run(
    originalDocument: Node,
    execution: ProcessorExecution,
    action: () => void
  ): DocumentProcessingResult {
    try {
      action();
    } catch (error) {
      if (error instanceof RunTerminationError) {
        return execution.result();
      }
      if (error instanceof MustUnderstandFailure) {
        const failureDocument = originalDocument.clone() as BlueNode;
        return DocumentProcessingResult.capabilityFailure(
          failureDocument,
          error.message ?? null
        );
      }
      throw error;
    }
    return execution.result();
  }

  private initializationMarker(document: Node): BlueNode | null {
    const contracts = document.getProperties()?.contracts;
    const marker = contracts?.getProperties()?.initialized ?? null;
    if (!marker) {
      return null;
    }
    if (!(marker instanceof BlueNode)) {
      throw new ProcessorFatalError(
        'Initialization Marker must be a BlueNode',
        ProcessorErrors.illegalState('Initialization Marker must be a BlueNode')
      );
    }
    const typeBlueId = marker.getType()?.getBlueId();
    if (typeBlueId !== 'InitializationMarker') {
      throw new ProcessorFatalError(
        "Initialization Marker must declare type 'InitializationMarker'",
        ProcessorErrors.illegalState(
          "Initialization Marker must declare type 'InitializationMarker'"
        )
      );
    }
    return marker;
  }

  static nodeAt(root: Node, pointer: string): Node | null {
    if (!(root instanceof BlueNode)) {
      return null;
    }
    const normalized = normalizePointer(pointer);
    if (normalized === '/') {
      return root;
    }
    const segments = normalized.slice(1).split('/');
    let current: Node | null = root;

    for (const segment of segments) {
      if (!current) {
        return null;
      }
      if (!segment) {
        continue;
      }

      const items = current.getItems();
      if (items && /^\d+$/.test(segment)) {
        const index = Number.parseInt(segment, 10);
        current = items[index] ?? null;
        continue;
      }

      const properties = current.getProperties() as
        | Record<string, Node>
        | undefined;
      if (!properties) {
        return null;
      }
      const nextNode = properties[segment] as Node | undefined;
      if (!(nextNode instanceof BlueNode)) {
        return null;
      }
      current = nextNode;
    }

    return current;
  }
}

File: references/java-sdk/src/main/java/blue/language/processor/ProcessorEngine.java
package blue.language.processor;

import blue.language.model.Node;
import blue.language.processor.model.ChannelContract;
import blue.language.processor.model.Contract;
import blue.language.processor.model.HandlerContract;
import blue.language.processor.model.JsonPatch;
import blue.language.processor.util.NodeCanonicalizer;
import blue.language.processor.util.PointerUtils;
import blue.language.processor.util.ProcessorContractConstants;
import blue.language.processor.util.ProcessorPointerConstants;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

final class ProcessorEngine {

    private ProcessorEngine() {
    }

    static DocumentProcessingResult initializeDocument(DocumentProcessor owner, Node document) {
        Objects.requireNonNull(document, "document");
        if (isInitialized(owner, document)) {
            throw new IllegalStateException("Document already initialized");
        }
        Execution execution = new Execution(owner, document.clone());
        try {
            execution.initializeScope("/", true);
        } catch (RunTerminationException ignored) {
            // Initialization run terminated early (e.g., graceful root termination).
        } catch (MustUnderstandFailureException ex) {
            return DocumentProcessingResult.capabilityFailure(document.clone(), ex.getMessage());
        }
        return execution.result();
    }

    static DocumentProcessingResult processDocument(DocumentProcessor owner, Node document, Node event) {
        Objects.requireNonNull(document, "document");
        Objects.requireNonNull(event, "event");
        if (!isInitialized(owner, document)) {
            throw new IllegalStateException("Document not initialized");
        }
        Node cloned = document.clone();
        Execution execution = new Execution(owner, cloned);
        try {
            execution.loadBundles("/");
            execution.processExternalEvent("/", event);
        } catch (RunTerminationException ignored) {
            // Processing terminated early; result still returned.
        } catch (MustUnderstandFailureException ex) {
            return DocumentProcessingResult.capabilityFailure(document.clone(), ex.getMessage());
        }
        return execution.result();
    }

    static boolean isInitialized(DocumentProcessor owner, Node document) {
        Objects.requireNonNull(document, "document");
        String pointer = PointerUtils.resolvePointer("/", ProcessorPointerConstants.RELATIVE_INITIALIZED);
        Node marker = null;
        try {
            marker = nodeAt(document, pointer);
        } catch (Exception ignored) {
        }
        if (marker == null) {
            return false;
        }
        validateInitializationMarker(marker, pointer);
        return true;
    }

    @SuppressWarnings("unchecked")
    static ChannelMatch evaluateChannel(DocumentProcessor owner,
                                            String bindingKey,
                                            ChannelContract contract,
                                            ContractBundle bundle,
                                            String scopePath,
                                            Node event) {
        ChannelProcessor<? extends ChannelContract> processor = owner.registry().lookupChannel(contract).orElse(null);
        if (processor == null) {
            return ChannelMatch.noMatch();
        }
        Node clonedEvent = event != null ? event.clone() : null;
        Object eventObject = null;
        try {
            eventObject = owner.contractConverter().convertWithType(clonedEvent, Object.class, false);
        } catch (Exception ignored) {
        }
        @SuppressWarnings("unchecked")
        ChannelProcessor<ChannelContract> typed = (ChannelProcessor<ChannelContract>) processor;
        ChannelEvaluationContext context = new ChannelEvaluationContext(scopePath,
                bindingKey,
                clonedEvent,
                eventObject,
                bundle.markers(),
                bundle,
                owner.registry());
        ChannelProcessorEvaluation evaluation = typed.evaluate(contract, context);
        if (evaluation == null || !evaluation.matches()) {
            return ChannelMatch.noMatch();
        }
        Node eventNode = evaluation.eventNode() != null ? evaluation.eventNode() : context.event();
        return new ChannelMatch(true, evaluation.eventId(), typed, context, eventNode, evaluation.deliveries());
    }

    static Node createLifecycleInitiatedEvent(String documentId) {
        Node event = new Node()
                .type(new Node().blueId("Document Processing Initiated"))
                .properties("type", new Node()
                .value("Document Processing Initiated")
                .type(new Node().blueId("Document Processing Initiated")));
        event.properties("documentId", new Node().value(documentId));
        return event;
    }

    static String canonicalSignature(Node node) {
        return NodeCanonicalizer.canonicalSignature(node);
    }

    static Node createDocumentUpdateEvent(DocumentProcessingRuntime.DocumentUpdateData data, String scopePath) {
        String relativePath = PointerUtils.relativizePointer(scopePath, data.path());
        Node event = new Node().properties("type", new Node().value("Document Update"));
        event.properties("op", new Node().value(data.op().name().toLowerCase()));
        Node beforeNode = data.before() != null ? data.before().clone() : new Node().value(null);
        Node afterNode = data.after() != null ? data.after().clone() : new Node().value(null);
        event.properties("path", new Node().value(relativePath));
        event.properties("before", beforeNode);
        event.properties("after", afterNode);
        return event;
    }

    static boolean matchesDocumentUpdate(String scopePath, String watchPath, String changedPath) {
        if (watchPath == null || watchPath.isEmpty()) {
            return false;
        }
        String watch = PointerUtils.resolvePointer(scopePath, watchPath);
        String changed = PointerUtils.normalizePointer(changedPath);
        if (watch.equals("/")) {
            return true;
        }
        if (changed.equals(watch)) {
            return true;
        }
        return changed.startsWith(watch + "/");
    }

    static Node nodeAt(Node root, String pointer) {
        Objects.requireNonNull(root, "root");
        String[] segments = PointerUtils.splitPointerSegments(pointer);
        Node current = root;
        for (String segment : segments) {
            if (current == null) {
                return null;
            }
            Map<String, Node> props = current.getProperties();
            if (props != null && props.containsKey(segment)) {
                current = props.get(segment);
                continue;
            }

            if (PointerUtils.isArrayIndexSegment(segment)) {
                int index = PointerUtils.parseArrayIndex(segment);
                List<Node> items = current.getItems();
                if (items == null || index < 0 || index >= items.size()) {
                    return null;
                }
                current = items.get(index);
                continue;
            }

            if ("type".equals(segment)) {
                current = current.getType();
                continue;
            }
            if ("itemType".equals(segment)) {
                current = current.getItemType();
                continue;
            }
            if ("keyType".equals(segment)) {
                current = current.getKeyType();
                continue;
            }
            if ("valueType".equals(segment)) {
                current = current.getValueType();
                continue;
            }
            if ("blue".equals(segment)) {
                current = current.getBlue();
                continue;
            }

            return null;
        }
        return current;
    }

    static boolean hasInitializationMarker(Node root, String scopePath) {
        String pointer = PointerUtils.resolvePointer(scopePath, ProcessorPointerConstants.RELATIVE_INITIALIZED);
        Node marker;
        try {
            marker = nodeAt(root, pointer);
        } catch (Exception ignored) {
            return false;
        }
        if (marker == null) {
            return false;
        }
        validateInitializationMarker(marker, pointer);
        return true;
    }

    static void validateInitializationMarker(Node marker, String pointer) {
        if (marker == null) {
            return;
        }
        Node type = marker.getType();
        String blueId = type != null ? type.getBlueId() : null;
        boolean validBlueId = "EVguxFmq5iFtMZaBQgHfjWDojaoesQ1vEXCQFZ59yL28".equals(blueId)
                || "Processing Initialized Marker".equals(blueId)
                || "Core/Processing Initialized Marker".equals(blueId)
                || "InitializationMarker".equals(blueId);
        if (!validBlueId) {
            throw new IllegalStateException(
                    "Reserved key 'initialized' must contain an Initialization Marker at " + pointer);
        }
    }

    static final class Execution {
        private final DocumentProcessor owner;
        private final DocumentProcessingRuntime runtime;
        private final Map<String, ContractBundle> bundles = new LinkedHashMap<>();
        private final Map<String, PendingTermination> pendingTerminations = new LinkedHashMap<>();
        private final Set<String> cutOffScopes = new LinkedHashSet<>();
        private final CheckpointManager checkpointManager;
        private final TerminationService terminationService;
        private final ChannelRunner channelRunner;
        private final ScopeExecutor scopeExecutor;

        Execution(DocumentProcessor owner, Node document) {
            this.owner = owner;
            this.runtime = new DocumentProcessingRuntime(document);
            this.checkpointManager = new CheckpointManager(runtime, ProcessorEngine::canonicalSignature);
            this.terminationService = new TerminationService(runtime);
            this.channelRunner = new ChannelRunner(owner, this, runtime, checkpointManager);
            this.scopeExecutor = new ScopeExecutor(owner, this, runtime, bundles, channelRunner);
        }

        void initializeScope(String scopePath, boolean chargeScopeEntry) {
            scopeExecutor.initializeScope(scopePath, chargeScopeEntry);
        }

        void loadBundles(String scopePath) {
            scopeExecutor.loadBundles(scopePath);
        }

        void processExternalEvent(String scopePath, Node event) {
            scopeExecutor.processExternalEvent(scopePath, event);
        }

        void handlePatch(String scopePath,
                         ContractBundle bundle,
                         JsonPatch patch,
                         boolean allowReservedMutation) {
            handlePatch(scopePath, bundle, patch, allowReservedMutation, false);
        }

        void handlePatch(String scopePath,
                         ContractBundle bundle,
                         JsonPatch patch,
                         boolean allowReservedMutation,
                         boolean allowTerminatedWork) {
            scopeExecutor.handlePatch(scopePath, bundle, patch, allowReservedMutation, allowTerminatedWork);
        }

        ProcessorExecutionContext createContext(String scopePath,
                                                ContractBundle bundle,
                                                Node event) {
            return createContext(scopePath, bundle, event, false, false);
        }

        ProcessorExecutionContext createContext(String scopePath,
                                                ContractBundle bundle,
                                                Node event,
                                                boolean allowTerminatedWork) {
            return createContext(scopePath, bundle, event, allowTerminatedWork, false);
        }

        ProcessorExecutionContext createContext(String scopePath,
                                                ContractBundle bundle,
                                                Node event,
                                                boolean allowTerminatedWork,
                                                boolean allowReservedMutation) {
            return new ProcessorExecutionContext(this, bundle, scopePath,
                    cloneEvent(event), allowTerminatedWork, allowReservedMutation);
        }

        DocumentProcessingResult result() {
            return DocumentProcessingResult.of(runtime.document(), runtime.rootEmissions(), runtime.totalGas());
        }

        DocumentProcessingRuntime runtime() {
            return runtime;
        }

        DocumentProcessor owner() {
            return owner;
        }

        boolean isScopeInactive(String scopePath) {
            String normalized = PointerUtils.normalizeScope(scopePath);
            return cutOffScopes.contains(normalized)
                    || pendingTerminations.containsKey(normalized)
                    || runtime.isScopeTerminated(normalized);
        }

        void enterGracefulTermination(String scopePath, ContractBundle bundle, String reason) {
            terminate(scopePath, bundle, ScopeRuntimeContext.TerminationKind.GRACEFUL, reason);
        }

        void enterFatalTermination(String scopePath, ContractBundle bundle, String reason) {
            terminate(scopePath, bundle, ScopeRuntimeContext.TerminationKind.FATAL, reason);
        }

        private void terminate(String scopePath,
                               ContractBundle bundle,
                               ScopeRuntimeContext.TerminationKind kind,
                               String reason) {
            String normalized = PointerUtils.normalizeScope(scopePath);
            if (pendingTerminations.containsKey(normalized) || runtime.isScopeTerminated(normalized)) {
                return;
            }
            pendingTerminations.put(normalized, new PendingTermination(kind, reason));
            terminationService.terminateScope(this, scopePath, bundle, kind, reason);
        }

        ContractBundle bundleForScope(String scopePath) {
            return bundles.get(PointerUtils.normalizeScope(scopePath));
        }

        void recordPendingTermination(String scopePath,
                                      ScopeRuntimeContext.TerminationKind kind,
                                      String reason) {
            pendingTerminations.put(PointerUtils.normalizeScope(scopePath), new PendingTermination(kind, reason));
        }

        void clearPendingTermination(String scopePath) {
            pendingTerminations.remove(PointerUtils.normalizeScope(scopePath));
        }

        void markCutOff(String scopePath) {
            String normalized = PointerUtils.normalizeScope(scopePath);
            if (cutOffScopes.add(normalized)) {
                ScopeRuntimeContext context = runtime.existingScope(normalized);
                if (context != null) {
                    context.markCutOff();
                }
            }
        }

        String fatalReason(Throwable throwable, String defaultReason) {
            String message = throwable != null ? throwable.getMessage() : null;
            return message != null ? message : defaultReason;
        }

        void deliverLifecycle(String scopePath,
                              ContractBundle bundle,
                              Node event,
                              boolean finalizeAfter) {
            scopeExecutor.deliverLifecycle(scopePath, bundle, event, finalizeAfter);
        }

        void recordLifecycleForBridging(String scopePath, Node event) {
            ScopeRuntimeContext scopeContext = runtime.scope(scopePath);
            scopeContext.recordBridgeable(event.clone());
            if ("/".equals(scopePath)) {
                runtime.recordRootEmission(event.clone());
            }
        }

        private Node cloneEvent(Node event) {
            return event != null ? event.clone() : null;
        }
    }

    private static final class PendingTermination {
        final ScopeRuntimeContext.TerminationKind kind;
        final String reason;

        PendingTermination(ScopeRuntimeContext.TerminationKind kind, String reason) {
            this.kind = kind;
            this.reason = reason;
        }
    }


    @SuppressWarnings("unchecked")
    static void executeHandler(DocumentProcessor owner, HandlerContract contract, ProcessorExecutionContext context) {
        HandlerProcessor<? extends HandlerContract> processor = owner.registry()
                .lookupHandler(contract)
                .orElseThrow(() -> new IllegalStateException(
                        "No processor registered for contract type " + contract.getClass().getName()));
        HandlerProcessor<HandlerContract> typed = (HandlerProcessor<HandlerContract>) processor;
        if (!typed.matches(contract, context)) {
            return;
        }
        typed.execute(contract, context);
    }

    static final class ChannelMatch {
        final boolean matches;
        final String eventId;
        final ChannelProcessor<ChannelContract> processor;
        final ChannelEvaluationContext context;
        final Node eventNode;
        final List<ChannelProcessorEvaluation.ChannelDelivery> deliveries;

        ChannelMatch(boolean matches,
                     String eventId,
                     ChannelProcessor<ChannelContract> processor,
                     ChannelEvaluationContext context,
                     Node eventNode,
                     List<ChannelProcessorEvaluation.ChannelDelivery> deliveries) {
            this.matches = matches;
            this.eventId = eventId;
            this.processor = processor;
            this.context = context;
            this.eventNode = eventNode;
            this.deliveries = deliveries;
        }

        Node eventNode() {
            return eventNode;
        }

        static ChannelMatch noMatch() {
            return new ChannelMatch(false, null, null, null, null, java.util.Collections.<ChannelProcessorEvaluation.ChannelDelivery>emptyList());
        }
    }

    static final class BoundaryViolationException extends RuntimeException {
        BoundaryViolationException(String message) {
            super(message);
        }
    }
}
